bundles/org.platformio.eclipse.ide.home.ui/src/org/platformio/eclipse/ide/home/internal/ui/handlers/PlatformIOHandler.java

```diff
@@ -20,37 +20,55 @@
  *******************************************************************************/
 package org.platformio.eclipse.ide.home.internal.ui.handlers;
 
+import java.io.IOException;
 import java.util.Optional;
+import java.util.stream.Stream;
 
 import org.eclipse.core.commands.AbstractHandler;
 import org.eclipse.core.commands.ExecutionEvent;
 import org.eclipse.core.commands.ExecutionException;
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.runtime.CoreException;
-import org.eclipse.core.runtime.Platform;
 import org.eclipse.core.runtime.ServiceCaller;
+import org.eclipse.core.runtime.jobs.Job;
 import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.ui.console.ConsolePlugin;
+import org.eclipse.ui.console.IOConsoleOutputStream;
 import org.eclipse.ui.handlers.HandlerUtil;
+import org.platformio.eclipse.ide.home.api.Input;
+import org.platformio.eclipse.ide.home.api.Output;
 import org.platformio.eclipse.ide.home.api.PlatformIO;
+import org.platformio.eclipse.ide.home.core.DefaultInput;
+import org.platformio.eclipse.ide.home.internal.ui.terminal.Terminal;
+import org.platformio.eclipse.ide.home.internal.ui.terminal.TerminalFactory;
+import org.platformio.eclipse.ide.home.internal.ui.terminal.TerminalOutput;
 
 public abstract class PlatformIOHandler extends AbstractHandler {
 
 	@Override
 	public final Object execute(ExecutionEvent event) throws ExecutionException {
 		IStructuredSelection selection = HandlerUtil.getCurrentStructuredSelection(event);
 		Optional<IProject> project = new SelectProject(selection).get();
+		Optional<Terminal> terminal = Stream.of(ConsolePlugin.getDefault().getConsoleManager().getConsoles())
```

Comment:
There should be an utility like `EnsureOpenTerminal` that always return `Terminal` of fail